### DISCUSSION TITLE
Modifying existing Web Radio URL is not saved to Shared data (NFS other server)

### DISCUSSION THREAD

**Discussion by Sagnasty** [rern/rAudio#1253] [Fixed Issues]
Running rAudio 20230224

To reproduce problem:
- Modify existing Web Radio URL and save

View web radio URL on another Raudio machine  (windows / firefox browser) - webradio URL remains unchanged.

Rebooting raudio systems has no affect.

All other shared data (NAS) seems to update as normal.

Jim

3 x Raspberry Pi 3B+ Rev 1.3 > HiFiBerry DAC+ Pro
1 x Raspberry Pi 3B Rev 1.2 > HiFiBerry DAC+ Pro
Synology DS718+

Jim

**Comment by Sagnasty** [comment]
Another note: 
For a quick test, I created a new web radio station.  
The web radio count was automatically incremented correctly on the other Raudio machines, but the new web radio information was not.

**Comment by ceejayemm** [comment]
Jim

Check what permissions / owners you have on the actual NFS share. Rern assumes that the NFS share has 'his' required permissions and sometimes tries to change them if not - which is not allowed, and probably won't work on the share, the permissions and ownership have to be set and maintained by the system (server - in your case the Synology) providing the NFS share. I have had similar problems with  NFS shares and ended up with some really odd combinations of permissions, owners etc. 

The problem with NFS is knowing which (remote) user is accessing the share and translating that user into something the NFS server knows about and can thus apply the correct permissions to the request. You can have hours or fun with this as I have found out in the past. In some ways using CIFS (I hate to say) makes life easier as you can explicitly set the user/password needed to access the share and achieve the required access permissions. 

When using NFS shares with rAudio, it might be best to give everybody full access to everything on the NFS share - not entirely safe (!) but at least it generally works.  Coming from an enterprise background where security is paramount this gives me the shakes :-)

All the above is based on my experiences with OpenMediaVault v6.x and so might be different to your Synology NAS setup. which I know nothing about. They may, or may not, however be useful.

Regards.

Chris

**Comment by Sagnasty** [comment]
Chris,

My NFS share permissions have not changed for quite some time now and I believe this is the first time I have modified data on an existing web radio with the new Shared data implementation.   

One thing I did notice with shared data DISabled is that the file permissions / groups are different in the following way (determined by comparing web radio information housed in the backup \data\webradio\ folder taken from the machine that the web radio updates where originally made):
UNmodified web radio(s):  Mode - rwx r-x r-x   User & Group http
Modified web radio(s):       Mode - rw- r-- r--    User & Group root
New web radio:                  Mode - rw- r-- r--    User & Group http

Note: the UNmodfied web radios were created prior to the new Shared data implementation.


Another oddity: all web radio information updated correctly on my secondary machine once I Re-ENabled shared data on the machine that I originally performed the web radio updates to.     

The shared data information for Web radios on the NAS (before and after enabling shared data) is identical.   This seems to imply that the web radio information update is machine to machine instead of machine to/from NAS server.   If this is not the case, then where is the web radio information saved in the shared data file structure?

**Comment by ceejayemm** [comment]
Jim

Also check what the permissions of the files / folders are on the Synology itself (not the rAudio end but on the server itself).  I have seen many and various owners/permissions at that end - including user/group 'nobody'.  I found that I could write the initial setup of the SharedData folders to the NFS share but then not be able to change them at a later date. It would appear that the create stage potentially runs as one user and then later on the edit/modify process runs as a different user.

I have found that, certainly with the later versions of SharedData, that the following folders are copied to the share (then referenced as symlinks):

lrwxrwxrwx 1 http http    25 Feb 25 10:06 audiocd -> /mnt/MPD/NAS/data/audiocd
lrwxrwxrwx 1 http http    27 Feb 25 10:06 bookmarks -> /mnt/MPD/NAS/data/bookmarks
lrwxrwxrwx 1 http http    24 Feb 25 10:06 lyrics -> /mnt/MPD/NAS/data/lyrics
lrwxrwxrwx 1 mpd  audio   21 Feb 25 10:06 mpd -> /mnt/MPD/NAS/data/mpd
lrwxrwxrwx 1 mpd  audio   27 Feb 25 10:06 playlists -> /mnt/MPD/NAS/data/playlists
lrwxrwxrwx 1 http http    26 Feb 25 10:06 webradio -> /mnt/MPD/NAS/data/webradio

As you can see, the webradio folder IS copied to the SharedData location. So all clients using the SharedData folders will all see the same set of radio stations/images BUT each can play its own selection. As can any item from any of these folders. You can also see that already some folders have different owners/groups (mpd and playlists).

As noted previously I have now moved to CIFS shares (on OMV) for SharedData rather than NFS shares and so far, touch wood, I haven't had any further failures to update these shared items as needed. After creating the CIFS share, the first client that connects to SharedData gives the option to copy the local folders to the share.  Subsequent connections then all see, and connect to, the same set of folders/files AND can make any relevant changes to them.  As also previously noted, I would prefer to use NFS shares, this being Linux and all, but on this occasion the CIFS variation actually seems to work better and in a manner that makes sense to me. I do however still use NFS shares from my music library but in light of the above I might consider changing to CIFS at some point for this too.

My workflow to move to CIFS for SharedData was as  follows:

1. Disable SharedData on all rAudio clients (at this point I was using rAudio NFS shares, not OMV NFS shares).
2. Create a new (empty) CIFS share, in my case on OMV, and assign user credentials for access control.
3. On one rAudio instance, make sure the Library and all other data areas are date as you want them.
4. On this same rAudio instance, enable SharedData and configure access to the new CIFS share and credentials as created in (2) above.
5. Take the option to copy the local data to the new CIFS share.
6. Once the copy is complete check the Library to make sure that all the same items are present and accessible.
7. Configure SharedData on each remaining rAudio clients using the same CIFS share and credentials as created in (2) above.
8. Check the Library on each newly configured rAudio Library to make sure you have what you want before moving on to the next client.

I hope this helps.

Chris

**Reply by Sagnasty** [reply]
Hello,

For sanity’s sake (based on the strange web radio file permissions observed from the backup), I decided to perform a fresh test starting with the latest image (rAudio-64bit-20230212.img)  on my Pi 3B+ Rev 1.3. 
Applied the latest updates (rAudio 20230224).
Manually re-configured my system from default (i.e., not restored from backup).
Confirmed that no other raudio machine was turned on.
Deleted the entire sharedata file structure from my NAS.
Confirmed that the NFS share was configured with full rwx permissions for everyone (Note: My sharedata file structure has the exact same properties as my music file structure.  These permissions apply to all folders, subfolders, and files.   Please note, there have been no changes to the NAS NFS settings for at least a year now).
Successfully re-entered the NFS network storage for my music files and refreshed the database.

I elected to work with the four radio paradise web radios for my initial test to keep the system as standard as possible.
The initial library info (local data):
<img src="https://user-images.githubusercontent.com/76575081/222172699-2db08efc-3e5e-48eb-bfaa-6c606d54da1e.png" width=300>
<img src="https://user-images.githubusercontent.com/76575081/222172738-8af74c53-1c4d-436b-b20a-f6e95f007348.png" width=300>
   
Saved a backup (No shared data).
Enabled NFS sharedata.
The following structure was immediately created on my NAS (as view from Synology file station):
<img src="https://user-images.githubusercontent.com/76575081/222172822-27cc3fec-e962-4fc7-a091-7c52ab6401ef.png" width=300>

Note that even the “webradio” directory was not created nor was an error generated.  

The resulting library info: 
<img src="https://user-images.githubusercontent.com/76575081/222174514-9daa097f-d4ed-4bbd-8b8c-9ea494eb2a9c.png" width=300>
<img src="https://user-images.githubusercontent.com/76575081/222173003-2ec09a5b-d20a-4c6a-8948-e43cde4632e1.png" width=300>

Saved a backup (shared data enabled).

The good news is that I can repeatably enable and disable sharedata and reach the same end result as described above for each scenario.

Jim

**Comment by ceejayemm** [comment]
Jim

Strange!  The local folder /srv/http/data/webradio was simlinked to /mnt/MPD/NAS/data/webradio the first time I accessed the SharedData CIFS folder.  I did however already have 24 stations in the local folder including the default Paradise Radio stations and they were all copied to the SharedData location.

Chris

**Comment by Sagnasty** [comment]
Chris

Agree.  I really did not expect that result - I have been using the new version of NFS shared data since late 2022 and I have never seen this problem.

Also, as a test,  I copied (via SFTP) the webradio and image files from this machine to the sharedata webradio folder on the NAS.   After enabling NFS shared data, all four radio paradise web radio links were present in the library view and could be played without issue.

Jim

**Comment by Sagnasty** [comment]
Hello,

Current Shared data (NFS) behavior with rAudio 20230303:

Confirmed that no other raudio machine was turned on.
Manually remove all data from NAS sharedata folder.
Ensure that shared data is disabled.


Enable Shared Data (NFS).
After about 5 seconds, the Mount to NAS shareddata foleder is successfully created and the Shared Data client slider control automatically moves from right to left (off).
No data is created on the NAS sharedata folder.

Also, If you "forget" the mount to shareddata folder, the system will immediately begin to update the library database after successfully removing this mount.

Jim

**Comment by ceejayemm** [comment]
This gets weirder.

I had to reflash my SD card with i20230212 (RPi 3A+ 64 bit) and then applied the latest update 20230304. I reconfigured everything from scratch and then enabled SharedDate from my existing CIFS share (which has work flawlessly up to now - as noted above).  Now the System page shows the CIFS share as mounted  

{
	"audio": true,
	"audioaplayname": "",
	"audiocards": false,
	"audiooutput": "",
	"bluetooth": true,
	"bluetoothactive": true,
	"bluetoothconf": [
		false,
		false
	],
	"btconnected": false,
	"display": {
		"logout": false
	},
	"hddapm": false,
	"hddsleep": false,
	"hdmi": false,
	"hostapd": false,
	"hostname": "rAudio-1",
	"i2seeprom": false,
	"i2smodulesw": false,
	"lcd": false,
	"lcdchar": false,
	"lcdcharaddr": false,
	"lcdcharconf": false,
	"lcdcharreboot": false,
	"lcdmodel": "",
	"lcdreboot": false,
	"list": [
		{
			"icon": "microsd",
			"mountpoint": "/mnt/MPD/SD",
			"mounted": true,
			"source": "/dev/mmcblk0p2",
			"size": "2.8GB/29GB",
			"nfs": false,
			"smb": false
		},
		{
			"icon": "networks",
			"mountpoint": "/mnt/MPD/NAS/data",
			"mounted": true,
			"source": "//192.168.86.250/raudio",
			"size": "994GB/3.6TB"
		},
		{
			"icon": "networks",
			"mountpoint": "/mnt/MPD/NAS/NAS_Music",
			"mounted": true,
			"source": "192.168.86.250:/export/music/iTunes/Music",
			"size": "994GB/3.6TB"
		}
	],
	"login": false,
	"mpdoled": false,
	"mpdoledconf": [
		6,
		800000
	],
	"mpdoledreboot": false,
	"nfsserver": false,
	"page": "system",
	"powerbutton": false,
	"powerbuttonconf": false,
	"relays": false,
	"rotaryencoder": false,
	"rotaryencoderconf": false,
	"shareddata": false,
	"soundprofile": false,
	"soundprofileconf": [],
	"status": "0.00 • 0.02 • 0.03
45.1 °C
2023-03-07 • 18:30:12 Europe · London
18m since 2023-03-07 • 18:11
",
	"system": "rAudio 20230304
6.1.11-1 aarch64
Raspberry Pi 3A+ Rev 1.0
BCM2837B0 • 449 MiB
4 x Cortex-A53 @ 1.40 GHz",
	"timezone": "Europe/London",
	"timezoneoffset": "+00:00",
	"usbautoupdate": false,
	"vuled": false,
	"vuledconf": false,
	"warning": "",
	"wlan": true,
	"wlanconf": [
		"00",
		false
	],
	"wlanconnected": true
}

and is accessible via an SSH terminal session:

rAudio-1:/srv/http/data  #  ls -lrt /mnt/MPD/NAS/data         
total 4
drwxr-xr-x 2 mpd mpd  0 Feb 13 09:34 audiocd
drwxr-xr-x 2 mpd mpd  0 Feb 13 09:34 lyrics
drwxr-xr-x 2 mpd mpd  0 Feb 13 09:34 playlists
drwxr-xr-x 2 mpd mpd  0 Feb 13 09:34 system
drwxr-xr-x 2 mpd mpd  0 Feb 25 10:10 webradio
drwxr-xr-x 2 mpd mpd  0 Feb 25 10:11 bookmarks
drwxr-xr-x 2 mpd mpd  0 Feb 26 09:57 mpd
-rwxr-xr-x 1 mpd mpd 28 Mar  7 18:11 sharedip

However the SharedData setup is NOT available for this system to use, only the local file system which currently empty (other than the Paradise Radio stations delivered as part of the img):

rAudio-1:/srv/http/data  #  ls -lrt
total 40
drwxr-xr-x 4 http http  4096 Jan 22 13:40 camilladsp
drwxr-xr-x 2 http http  4096 Jan 22 13:41 lyrics
drwxr-xr-x 2 http http  4096 Jan 22 13:41 bookmarks
drwxr-xr-x 2 http http  4096 Jan 22 13:41 audiocd
lrwxrwxrwx 1 http http     8 Jan 22 13:41 shm -> /dev/shm
drwxr-xr-x 2 mpd  audio 4096 Jan 22 13:54 playlists
drwxr-xr-x 3 http http  4096 Feb 12 07:56 webradio
drwxr-xr-x 2 mpd  audio 4096 Mar  7 17:18 mpd
drwxr-xr-x 2 http http  4096 Mar  7 18:05 addons
drwxr-xr-x 3 http http  4096 Mar  7 18:11 mpdconf
drwxr-xr-x 2 http http  4096 Mar  7 18:16 system


@rern - what is going on here and how do I get back to a working system using SharedData ?

Chris

**Reply by Sagnasty** [reply]
One thing I noticed after re-flashing all my machines to i20230212 (RPi 3A+ 64 bit) and latest update 20230304 and re-configuring everything from scratch is that I had to increase my mass_output_buffer size (Player / Buffer-Output).   

I was getting very strange behavior in my resulting library with No sharedata enabled until I increased the buffer size.   Switching between sharedata enabled and disabled produced even stranger behavior as one would expect.


My current system behavior as of today:

Confirmed that no other raudio machine was turned on.
Manually remove all data from NAS sharedata folder.
Ensure that shared data is disabled.

Enable Shared Data (NFS).
After about 5 seconds, the Mount to NAS shareddata folder is successfully created and the Shared Data client slider control automatically moves from right to left (off).
No data is created on the NAS sharedata folder.

Also, If you "forget" the mount to shareddata folder, the system will immediately begin to update the library database after successfully removing this mount.


System data of my local machine with sharedata enabled (NFS)
[System data with Sharedata.txt](https://github.com/rern/rAudio-1/files/10913755/System.data.with.Sharedata.txt)

<img src="https://user-images.githubusercontent.com/76575081/223545731-568827d6-c1ce-421d-b33a-f4bb50bbd14d.png" width=900>

No data is written to the NAS sharedata folder, but my library view is the same  and is functional (including the custom webradios) **with and without sharedata enabled**.


<img src="https://user-images.githubusercontent.com/76575081/223548948-c81586e3-c138-43a5-be9e-e4738d960ee9.png" width=300>


Jim

**Comment by ceejayemm** [comment]
'Forgetting' the ShareData mount point does indeed kick off an automatic update of the local Library and at the end of this update I can see all the contents of my Library - 333 Albums and 5040 items in total.  As you can see my library is quite small so I have never had to resort to lifting the max_output_buffer from its default setting.

Also noted that the System / Data and Settings / Shared Data setting does not stay enabled BUT the Shared Data CIFS folder specified IS mounted and, if accessed via SSH, does allow access to the data previously uploaded there by other RA clients.

@rern - This is now getting very annoying, please just let it work like it should !

Chris

**Comment by ceejayemm** [comment]
Ok, an update on this problem.

I flashed a new SD card using the i20230212 64 bit image to a RPi 3A+ and configured it for use, including setting up the SharedData settings to my existing CIFS based share. Checking the Library I found that all my expected Library items were present including all my WebRadio stations.  I then applied the 20230304 update and found that the SharedData settings and setup was preserved and usable.  

This seems to show that the problem appears to be down to the sequence in which the updates are applied:

1. Base image, update, configuration - appears to not allow SharedData to be used properly
2. Base image, configuration, update - appears to allow SharedData to work properly

I hope this helps get to the bottom of this issue.

Chris

**Comment by Sagnasty** [comment]
Using Sequence 2 (Base image, configuration, update) that @ceejayemm described above,  I can also confirm the same end result with a NFS based share.   

I flashed a SD card using the i20230212 64 bit image to a RPi 3B+, configured it for use, and then set up the SharedData settings to my existing NFS based share folder (empty).

The sharedata mount was created, and the shared data slider control stayed on the right.  

I confirmed that the sharedata data was correctly written to my NFS sharedata folder on my NAS and that the resulting library contained all expected items including my added WebRadio stations.

I then applied the latest update (20230304) and confirmed that the SharedData settings and setup was preserved and functional.

Jim

**Comment by Sagnasty** [comment]
No longer an issue with 20231022
